components/camel-git/src/main/java/org/apache/camel/component/RepositoryFactory.java

```diff
@@ -0,0 +1,111 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.camel.component;
+
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.net.URL;
+import java.nio.channels.Channels;
+import java.nio.channels.ReadableByteChannel;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Arrays;
+import java.util.List;
+
+import org.apache.camel.RuntimeCamelException;
+import org.apache.camel.component.git.GitEndpoint;
+import org.apache.camel.support.ResourceHelper;
+import org.apache.camel.util.ObjectHelper;
+import org.eclipse.jgit.lib.Repository;
+import org.eclipse.jgit.storage.file.FileRepositoryBuilder;
+import org.eclipse.jgit.util.SystemReader;
+
+public abstract class RepositoryFactory {
+
+    private static final SystemReader DEFAULT_INSTANCE;
+    private static final List<String> VALID_SCHEMES = Arrays.asList("classpath:", "file:", "http:", "https:");
+
+    static {
+        DEFAULT_INSTANCE = SystemReader.getInstance();
+    }
+
+    private RepositoryFactory() {
+    }
+
+    public static Repository of(GitEndpoint endpoint) {
+        if (ObjectHelper.isNotEmpty(endpoint.getGitConfigFile())) {
+            return resolveConfigFile(endpoint, endpoint.getGitConfigFile());
+        }
+        return getRepository(endpoint, DEFAULT_INSTANCE);
+    }
+
+    private static Repository resolveConfigFile(GitEndpoint endpoint, String uri) {
+        if (ObjectHelper.isEmpty(uri)) {
+            throw new IllegalArgumentException("URI to git config file must be supplied");
+        }
+
+        if (!ResourceHelper.hasScheme(uri) || !VALID_SCHEMES.contains(ResourceHelper.getScheme(uri))) {
+            throw new IllegalArgumentException(
+                    "URI to git config file must have scheme:path pattern where scheme could be classpath, file, http or https");
+        }
+
+        String schema = ResourceHelper.getScheme(uri);
+        String path = uri.substring(schema.length());
+
+        File gitConfigFile;
+        if (ResourceHelper.isClasspathUri(uri)) {
+            gitConfigFile = new File(endpoint.getClass().getClassLoader().getResource(path).getFile());
+        } else if (ResourceHelper.isHttpUri(uri)) {
+            try {
+                gitConfigFile = getTempFileFromHttp(uri);
+            } catch (IOException e) {
+                throw new RuntimeCamelException(String.format("Something goes wrong when loading: %s", uri), e);
+            }
+        } else { //load from system
+            gitConfigFile = new File(path);
+            if (!gitConfigFile.exists() || gitConfigFile.isDirectory()) {
+                throw new IllegalArgumentException(String.format("Invalid file: %s", path));
```

Comment:
And based on the comment above, I'd reword the exception message to: `"The configuration file at %s is unreadable (either missing, lacking proper access permission os is not a regular file")`